server/bin/pbench-server-prep-shim-002.py

```diff
@@ -0,0 +1,331 @@
+        self.ndups = ndups
+        self.nerrs = nerrs
+
+
+def fetch_config_val(config, ef):
+
+    qdir = config.get("pbench-server", "pbench-quarantine-dir")
+    if not qdir:
+        ef.write("Failed: getconf.py pbench-quarantine-dir pbench-server\n")
+        return None, None
+
+    qdir = Path(qdir).resolve()
+    if not qdir.is_dir():
+        ef.write(f"Failed: {qdir} does not exist, or is not a directory\n")
+        return None, None
+
+    # we are explicitly handling version-002 data in this shim
+    receive_dir_prefix = config.get("pbench-server", "pbench-receive-dir-prefix")
+    if not receive_dir_prefix:
+        ef.write("Failed: getconf.py pbench-receive-dir-prefix pbench-server\n")
+        return None, None
+
+    receive_dir = Path(f"{receive_dir_prefix}-002").resolve()
+    if not receive_dir.is_dir():
+        ef.write(f"Failed: {receive_dir} does not exist, or is not a directory\n")
+        return None, None
+
+    return (qdir, receive_dir)
+
+
+def qdirs_check(qdir_val, qdir, logger):
+    try:
+        os.makedirs(qdir)
+    except FileExistsError:
+        # directory already exists, ignore
+        pass
+    except Exception:
+        logger.exception(
+            "os.mkdir: Unable to create {} destination directory: {}", qdir_val, qdir,
+        )
+        return None
+    return qdir
+
+
+def md5_check(tb, tbmd5, logger):
+    # read the md5sum from md5 file
+    try:
+        with tbmd5.open() as f:
+            archive_md5_hex_value = f.readline().split(" ")[0]
+    except Exception:
+        logger.exception("Quarantine: Could not read {}", tbmd5)
+
+    # get hex value of the tarball's md5sum
+    try:
+        archive_tar_hex_value = md5sum(tb)
+    except Exception:
+        logger.exception("Quarantine: Could not read {}", tb)
+
+    return (archive_md5_hex_value, archive_tar_hex_value)
+
+
+def process_tb(config, logger, receive_dir, qdir_md5, duplicates, errors, ef):
+
+    # Check for results that are ready for processing: version 002 agents
+    # upload the MD5 file as xxx.md5.check and they rename it to xxx.md5
+    # after they are done with MD5 checking so that's what we look for.
+    list_check = glob.glob(
+        os.path.join(receive_dir, "**", "*.tar.xz.md5"), recursive=True
+    )
+
+    archive = Path(config.ARCHIVE)
+    logger.info("{}", config.TS)
+    list_check.sort()
+    nstatus = ""
+
+    ntotal = ntbs = nerrs = nquarantined = ndups = 0
+
+    for tbmd5 in list_check:
+        ntotal += 1
+
+        # full pathname of tarball
+        tb = Path(tbmd5[0:-4])
+        tbmd5 = Path(tbmd5)
+
+        # directory
+        tbdir = tb.parent
+
+        # resultname: get the basename foo.tar.xz and then strip the .tar.xz
+        resultname = tb.name
+
+        controller = tbdir.name
+        dest = archive / controller
+
+        if (
+            (dest / resultname).is_file()
+            and (dest / tbmd5.name).is_file()
+            and (dest / resultname).exists()
```

Comment:
Are not this two `.exists` checks redundant?

Reply:
No, one of them is for tarball and another one is for .md5

Reply:
So the redundancy is not about the two `.exists()` checks for each file, but for having an `.is_file()` and a `.exists()` check for the same file.  How can you have `.is_file()` return `True` and `.exists()` return `False`?  A file can't be a file if it doesn't exist, right?

Seems like just checking that both are files also checks for existence.

Reply:
Ohh.. Ok